operator/pkg/clustermesh/client.go

```diff
@@ -0,0 +1,170 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package clustermesh
+
+import (
+	"context"
+
+	discovery "k8s.io/api/discovery/v1"
+	v1 "k8s.io/api/discovery/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/watch"
+	discoveryv1 "k8s.io/client-go/kubernetes/typed/discovery/v1"
+	mcsapiv1alpha1 "sigs.k8s.io/mcs-api/pkg/apis/v1alpha1"
+
+	k8sClient "github.com/cilium/cilium/pkg/k8s/client"
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+const (
+	meshRealServiceNameLabel = "mesh.cilium.io/real-service-name"
+)
+
+// meshClient override a few EndpointSlice methods to add/remove "hacks" on EndpointSlice.
+// This is necessary because the service informer passed to the controller will create
+// fake services with this form `$svcName-$clusterName` whereas the actual Service
+// doesn't contain the $clusterName. So this client is used to intercept
+// list/watch calls to also make sure that we fake the service name refererenced
+// by EndpoinSlices and will revert those changes at the create/update level to make
+// sure that the correct data is written to the Kubernetes API.
+type meshClient struct {
+	k8sClient.Clientset
+}
+
+func (c meshClient) DiscoveryV1() discoveryv1.DiscoveryV1Interface {
+	return meshClientDiscoveryV1{c.Clientset.DiscoveryV1()}
+}
+
+type meshClientDiscoveryV1 struct {
+	discoveryv1.DiscoveryV1Interface
+}
+
+func (c meshClientDiscoveryV1) EndpointSlices(namespace string) discoveryv1.EndpointSliceInterface {
+	return meshClientEndpointSlice{c.DiscoveryV1Interface.EndpointSlices(namespace)}
+}
+
+type meshClientEndpointSlice struct {
+	discoveryv1.EndpointSliceInterface
+}
+
+func isServiceOwnerReferrence(ownerReference metav1.OwnerReference) bool {
+	return ownerReference.APIVersion == "v1" && ownerReference.Kind == "Service"
+}
+
+// addEndpointSliceMeshHacks fakes that the parent service of the EndpointSlice
+// has the cluster name in its name by changing the owner reference and the label
+// pointing to the service name. It also make sure that there is a label `mesh.cilium.io/real-service-name`
+// to store the real service name before its modification.
+func addEndpointSliceMeshHacks(endpointSlice *v1.EndpointSlice) {
+	if endpointSlice == nil || endpointSlice.Labels[discovery.LabelManagedBy] != ciliumControllerName {
+		return
+	}
+
+	endpointSlice.Labels[meshRealServiceNameLabel] = endpointSlice.Labels[discovery.LabelServiceName]
```

Comment:
In a normal scenario I don't believe so, but that's a good point, I'm adding an extra check for the empty labels. As for the non existent `discovery.LabelServiceName` same as it won't correctly link to a service otherwise.